File: typescript/maroctax/src/ir.ts
import type { FiscalConfig } from './types.js';
import { roundMoney } from './utils.js';

/**
 * Annual gross income tax from progressive barème.
 * @param netTaxableAnnual - Annual net taxable income in MAD.
 * @param config - Fiscal configuration for the payroll year.
 */
export function calculateIrAnnual(
  netTaxableAnnual: number,
  config: FiscalConfig,
): number {
  const taxable = Math.max(0, netTaxableAnnual);

  for (const bracket of config.irBrackets) {
    if (taxable <= bracket.upTo) {
      return roundMoney(Math.max(0, taxable * bracket.rate - bracket.deduction));
    }
  }

  return 0;
}

/**
 * Monthly income tax withheld from salary.
 * @param netTaxableMonthly - Monthly net taxable income in MAD.
 * @param config - Fiscal configuration for the payroll year.
 */
export function calculateIrMonthly(
  netTaxableMonthly: number,
  config: FiscalConfig,
): number {
  const annual = calculateIrAnnual(netTaxableMonthly * 12, config);
  return roundMoney(annual / 12);
}

File: dart/maroctax/lib/src/ir.dart
import 'fiscal_config.dart';
import 'types.dart';
import 'utils.dart';

/// Annual gross income tax from progressive barème.
double calculateIrAnnual(double netTaxableAnnual, [FiscalConfig? config]) {
  final cfg = config ?? getFiscalConfig();
  final taxable = netTaxableAnnual < 0 ? 0 : netTaxableAnnual;

  for (final bracket in cfg.irBrackets) {
    if (taxable <= bracket.upTo) {
      final ir = taxable * bracket.rate - bracket.deduction;
      return roundMoney(ir < 0 ? 0 : ir);
    }
  }

  return 0;
}

/// Monthly income tax withheld from salary.
double calculateIrMonthly(double netTaxableMonthly, [FiscalConfig? config]) {
  final annual = calculateIrAnnual(netTaxableMonthly * 12, config);
  return roundMoney(annual / 12);
}
